### PR TITLE
Ensure library is rebuilt when header files are modified

### PR DESCRIPTION
**Pull request by andresag01**
This PR ensures that compiler generated dependency information is included as recipes in the library/Makefile so that the library is rebuilt when the header files are modified. This solution is inspired on the information at: http://make.mad-scientist.net/papers/advanced-auto-dependency-generation/ and http://scottmcpeak.com/autodepend/autodepend.html.

### CODE REVIEW

**Comment by andresag01** [comment]
This PR solves the main issue mentioned in the description. However, as mentioned in http://scottmcpeak.com/autodepend/autodepend.html, this mechanism has issues when source files are renamed. The easiest solution is to call `make clean`, but this is not seamless from the user point of view. The websites provide another way of solving this problem by using tools like `sed` and `fmt`, but these utilities are normally not available on Windows. We could translate this into perl. Please let me know you thoughts.

File: library/Makefile [hunk 1]
```diff
@@ -18,6 +18,9 @@ endif
 # To compile on Plan9:
 # CFLAGS += -D_BSD_EXTENSION
 
+# Generate dependency information on headers in GNU Make format
+DEPFLAGS = -MT $@ -MMD -MP -MF $*.d
```
Comment (andresag01) [inline library/Makefile]: These flags only work for GCC, there are similer ones for other compilers e.g. ARMCC and IAR

**Comment by NWilson** [comment]
By far the simplest solution is to use the recommended build method, CMake.

> The main systems used for development are CMake and yotta. Those systems are always complete and up-to-date. The others should reflect all changes present in the CMake and yotta build system, but some features are not ported there by default.

Adding deps support to the manually-generated Makefiles is probably not worth it — out of interest, do you have a reason to prefer them to the CMake-generated Makefiles?

**Comment by sbutcher-arm** [comment]
@andresag01 - It's a design aim to keep the Makefile as simple as simple as possible, and to use as few GNU specific features as possible to ensure maximum portability.

How portable are these extensions you've used? And how GNU specific are they?

**Comment by andresag01** [comment]
@sbutcher-arm: The additional compile flags to generate dependency makefiles work with GCC, clang and mingw. I have tried this in Linux and it works as expected. (I will try it in Windows as well). I think the changes will not work with armcc though, but I do not know whether we have ever tried compiling with armcc and makefiles anyways...

**Comment by sbutcher-arm** [comment]
@andresag01 - We work on a *lot* of platforms currently, and that's at least in part because we've kept the Makefile simple and avoided cleverer, platform specific or proprietary features.

Can you check this how well this works on a BSD? BSD's have (or had) their own make.

Also, what's the fix to `scripts/config.pl` for? It looks functionally equivalent, and apart from being prettier, and easier to read logically, I can't quite see your purpose. Would that be better off in it's own PR?

**Comment by andresag01** [comment]
@sbutcher-arm: Please find the answer to both your questions below.

_We work on a lot of platforms currently, and that's at least in part because we've kept the Makefile simple and avoided cleverer, platform specific or proprietary features.
Can you check this how well this works on a BSD? BSD's have (or had) their own make._
>> I understand your concern, but I do not think this will be a problem. If you take a close look at the change, you will see that the only additional make feature I am using is the `-include`, but I have not found any compatibility problems. Other than this, the only other features that I am using are from the compiler NOT from the make build system.

_Also, what's the fix to `scripts/config.pl` for? It looks functionally equivalent, and apart from being prettier, and easier to read logically, I can't quite see your purpose. Would that be better off in it's own PR?_
>> The problem is that the `get` command of that script actually changes the content of the file, which is counter intuitive. This has caused problems in the past in our make build system because of missing dependencies, something that is specially problematic when running parallel builds.

**Comment by hanno-arm** [comment]
@mpg @andresag01 @sbutcher-arm What's the status of this? I stumbled over that issue, too, and think it would be beneficial to have the default makefile notice changes to the headers.

**Comment by mpg** [comment]
IMO the goal of the GNU Makefiles is to be as simple and toolchain-agnostic as possible, as they provide a kind of common ground that many tool (even some proprietary IDEs) understand. People (including us) who need more feature (in particular, proper dependency handling) should use CMake.

I understand that's a trade-off and there are many reasonable places to draw the line. While I'm quite happy with the line where it is (except it should probably be documented), if other people have stronger opinions about this, I'm not opposed to changing it.

**Comment by andresag01** [comment]
@mpg: While I am in favour of being as portable as possible, I think that in the case of the build system we will always be incompatible with some compiler somewhere. For example, there is a PR to add support for IAR somewhere because our Makefile did not accommodate it. I personally always use Make and found out how easy it is to make a mistake because the library does not recompile when the headers change, so I thought I would add the change.

In any case, for the sake of portability, what I can do is make the variable `DEPFLAGS` to be assigned if is not already set, so that if you are compiling with something that is not compatible with GCC then you can overwrite it to be empty. What do you think?

**Comment by mpg** [comment]
I fully agree we'll always be incompatible with something somewhere (for one thing, we require GNU make), and as I said I think it's a continuum and it's about drawing a more-or-less arbitrary line somewhere in that continuum.

Since I'm happy using CMake for all things more complex than "make clean all", I'm biased towards not caring too much about putting more effort in the Makefile, but if the feature is optional as you suggest, I'm not opposed to it - especially since you already implemented it.

**Comment by gilles-peskine-arm** [comment]
I too wish our makefiles had correct dependencies on headers — it's especially annoying to modify `config.h` and end up with inconsistent binaries because the dependencies aren't declared.

I'm not sure if this PR is the best approach however. Dependencies don't change so often, so how about committing the dependency list to the repository? We already have a few committed generated files, such as `library/error.c`.

* In the makefiles, use an `include` directive to include the dependency list file. This doesn't require anything other than GNU make, which we require anyway.
* We have a script generate the dependency list. It uses a compiler with a Unix-like command line (`cc -E`).

A limitation of this approach is that the build dependencies depend on the configuration. However, I think that as long as we generate the dependency list file in a “full” config, the worst that will happen is that in a non-full config, there are a few spurious dependencies on header files that aren't actually used.

**Comment by andresag01** [comment]
@gilles-peskine-arm: I suppose that at the end of the day this is a matter of preference. I personally do not like having to worry about when to regenerate the dependency files, I just want to ensure that the makefiles do it for me and forget about it. However, there is also the concern that the dependency generation might not work with some compilers.

Given that this works with gcc, and armcc (need to check, but I am sure there are flags for dependency generation) I would not be too worried about this, and in any case I can just change the PR to make it easy to disable it if the compiler does not support it. The bottomline, we just have to make a decision on: Do we want to fix the dependencies for Makefiles? If yes, do we want to commit the dependencies or regenerate them automatically?

**Comment by daverodgman** [comment]
Closing this as we are tracking the issue in #5398

**Pull request closed without merging**